### ISSUE TITLE
Validation layers say Vulkano is submitting a single-use command buffer more than once

### ISSUE THREAD

**Issue by tolziplohu** [vulkano-rs/vulkano#1294]
I'm trying to upload data to an image that I'm also using for rendering. Ideally I shouldn't have to worry about sync since I'm only using one queue, but unless I signal a fence and wait for it after rendering and before uploading, I get an `AccessError: AlreadyInUse`. If I don't also wait for the GPU after uploading, the application seems to work as intended but the validation layer I have active complains:

```
VUID-vkQueueSubmit-pCommandBuffers-00071(ERROR / SPEC): msgNum: 0 - VkCommandBuffer 0x55df70cb2f80[] is already in use and is not marked for simultaneous use. The Vulkan spec states: If any element of the pCommandBuffers member of any element of pSubmits was not recorded with the VK_COMMAND_BUFFER_USAGE_SIMULTANEOUS_USE_BIT, it must not be in the pending state (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkQueueSubmit-pCommandBuffers-00071)
    Objects: 1
        [0] 0, type: 6, name: NULL
UNASSIGNED-CoreValidation-DrawState-CommandBufferSingleSubmitViolation(ERROR / SPEC): msgNum: 0 - VkCommandBuffer 0x55df70cb2f80[] was begun w/ VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT set, but has been submitted 0x2times.
    Objects: 1
        [0] 0, type: 6, name: NULL
```

Here's some pieces of my code, I'll try to make a minimum working example when I have time.

```rust
        // Draw
        let command_buffer =
            AutoCommandBufferBuilder::primary_one_time_submit(win.device(), win.queue.family())
                .unwrap()
                .begin_render_pass(frame.framebuffer, false, clear_values.clone())
                .unwrap()
                .draw(
                    pipeline.clone(),
                    &win.dynamic_state,
                    BufferlessVertices {
                        vertices: 4,
                        instances: 1,
                    },
                    desc.clone(),
                    pc,
                )
                .unwrap()
                .end_render_pass()
                .unwrap()
                .build()
                .unwrap();
        let f = future
            .join(frame.acquire)
            .then_execute(win.queue.clone(), command_buffer)
            .unwrap()
            .then_swapchain_present(win.queue.clone(), win.swapchain.clone(), frame.image_num)
            .then_signal_fence_and_flush();

        match f {
            Ok(f) => {
                future = Box::new(f) as Box<_>;
            }
            Err(vulkano::sync::FlushError::OutOfDate) => {
                recreate_swapchain = true;
                future = Box::new(vulkano::sync::now(win.device())) as Box<_>;
            }
            Err(err) => {
                // We'll keep going, it's probably not a big deal
                println!("{:?}", err);
                future = Box::new(vulkano::sync::now(win.device())) as Box<_>;
            }
        }

        // Another thread sends a list of blocks of the image to update
        if let Ok(update) = recv.try_recv() {
            // AccessError if I remove this
            future
                .then_signal_fence_and_flush()
                .unwrap()
                .wait(None)
                .unwrap();
            future = Box::new(vulkano::sync::now(win.device()));

            for (loc, data) in update.blocks {
                let b = block_buf.chunk(data).unwrap();
                let cmd = AutoCommandBufferBuilder::primary_one_time_submit(
                    win.device(),
                    win.queue.family(),
                )
                .unwrap()
                .copy_buffer_to_image_dimensions(
                    b,
                    image.clone(),
                    loc,
                    [16, 16, 16],
                    0,
                    1,
                    0,
                )
                .unwrap()
                .build()
                .unwrap();
                future = Box::new(future.then_execute(win.queue.clone(), cmd).unwrap());
            }

            // Validation layer complains if I remove this
            future
                .then_signal_fence_and_flush()
                .unwrap()
                .wait(None)
                .unwrap();
            future = Box::new(vulkano::sync::now(win.device()));
        }
```
I'm using Vulkano 0.16.0 on Arch Linux with a Radeon R9 270x with mesa's RADV.

EDIT: Here's a gist with a single file demonstrating the issue: https://gist.github.com/tolziplohu/132f2695431f09bb5e4530accdf7eac2

**Comment by AustinJ235** [comment]
I am not exactly sure what you are trying to accomplish here, but a swapchain has several images. I believe after ``then_swapchain_present`` you shouldn't be using the image anymore. I think you are trying to do partial updates on the swapchain image? I think this is where the present modes ``VK_PRESENT_MODE_SHARED_DEMAND_REFRESH_KHR`` and ``PRESENT_MODE_SHARED_CONTINUOUS_REFRESH_KHR`` may be used but those are not implemented at this time.

https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPresentModeKHR.html
https://github.com/vulkano-rs/vulkano/blob/fb3ef9d43e924f5bbff7dd9021544db4e8bd916d/vulkano/src/swapchain/capabilities.rs#L86

Right now what you will have to do is have a separate image separate from the swapchain so a ``StorageImage`` and update that and copy it on to the provided swapchain image every frame. You'll only be able to write to the image after read access is done in the swapchain. That is why you'll need the ``then_signal_fence_and_flush``.

**Comment by tolziplohu** [comment]
`image` up there isn't the swapchain image, it's a 3D StorageImage I'm using to store data for use in my fragment shader. My example isn't really clear, I'll make a smaller main.rs to demonstrate the problem.

**Comment by AustinJ235** [comment]
Was this resolved with the pull request mentioned?

**Comment by tolziplohu** [comment]
Not completely - I still get the AccessError if I don't flush before copying, but validation layers no longer complain if I don't flush after copying. So it's not entirely fixed, but it's better.

**Comment by funmaker** [comment]
I've encountered similar issue and I think the root cause is the same. `CommandBufferExecFuture::build_submission` returns `SubmitAnyBuilder::CommandBuffer` even if it was already flushed. If you try to execute a command buffer using already flushed future it's previously submitted command buffers still get added to the submission. Futures like `FenceSignalFuture` always return `SubmitAnyBuilder::Empty` so they do not exhibit this issue.

I've made minimal reproduction [funmaker/vulkano-issue-1294](https://github.com/funmaker/vulkano-issue-1294)
